cpp/arcticdb/version/version_map_entry.hpp

```diff
@@ -83,6 +80,63 @@ struct LoadParameter {
     }
 };
 
+
+inline bool is_undeleted_strategy_subset(const LoadStrategy& left, const LoadStrategy& right){
+    switch (left.load_type_) {
+        case LoadType::NOT_LOADED:
+            return true;
+        case LoadType::LOAD_LATEST:
+            // LOAD_LATEST is not a subset of LOAD_DOWNTO because LOAD_DOWNTO may not reach the latest undeleted version.
+            return right.load_type_ != LoadType::NOT_LOADED && right.load_type_ != LoadType::LOAD_DOWNTO;
+        case LoadType::LOAD_DOWNTO:
+            if (right.load_type_ == LoadType::LOAD_ALL) {
+                return true;
+            }
+            if (right.load_type_ == LoadType::LOAD_DOWNTO && ((left.load_until_version_.value()>=0) == (right.load_until_version_.value()>=0))) {
+                // Left is subset of right only when the [load_until]s have same sign and left's version is >= right's version
+                return left.load_until_version_.value() >= right.load_until_version_.value();
+            }
+            break;
+        case LoadType::LOAD_FROM_TIME:
+            if (right.load_type_ == LoadType::LOAD_ALL){
+                return true;
+            }
+            if (right.load_type_ == LoadType::LOAD_FROM_TIME){
+                return left.load_from_time_.value() >= right.load_from_time_.value();
+            }
+            break;
+        case LoadType::LOAD_ALL:
+            return right.load_type_ == LoadType::LOAD_ALL;
+        default:
+            util::raise_rte("Invalid load type: {}", left.load_type_);
+    }
+    return false;
+}
+
+// Returns a strategy which is guaranteed to load all versions requested by left and right.
+// Works only on strategies with include_deleted=false.
+inline LoadStrategy union_of_undeleted_strategies(const LoadStrategy& left, const LoadStrategy& right){
+    util::check(!left.include_deleted_, "Trying to produce a union of undeleted strategies but left strategy includes deleted.");
+    util::check(!right.include_deleted_, "Trying to produce a union of undeleted strategies but right strategy includes deleted.");
+    if (is_undeleted_strategy_subset(left, right)){
+        return right;
+    }
+    if (is_undeleted_strategy_subset(right, left)){
+        return left;
+    }
+    // If none is subset of the other then we should load all versions;
+    return LoadStrategy{LoadType::LOAD_ALL, false};
```

Comment:
I'm wary of performance regressions with this, was the behaviour as conservative before?

Reply:
Updated the comment.

I think we cannot do better than this because the `LoadType`s are such that when none is subset of the other there isn't really a better thing to do than to LOAD_ALL.

Also yes, that's what the code was doing before. Code responsible for this checks used to be [here](https://github.com/man-group/ArcticDB/blob/a2aa6864e5eb65a779363aee2f924e0860a21576/cpp/arcticdb/version/version_map_batch_methods.cpp#L22-L80).